forem/forem
# Save Data Update Script errors to database and show them on http://localhost:3000/admin/data_update_scripts

## What type of PR is this? (check all applicable)


## Description

We have a data update scripts page thats a work in progress - http://localhost:3000/admin/data_update_scripts. On this page, we want to show the error when a data update script fails. 

As per the discussion on the issue, the way that we implement this is adding an error field to DataUpdateScript.

>Currently, the only way to view an error is via Honeybadger. I think storing the error on the DataUpdateScript would be >advantageous if it fails so we can view it in admin. I could see adding the error field and displaying it as a following up PR to >the initial "add this to admin" PR

Note: In the next PR, we will give the user the ability to re run the script and clear the error if it passes. 
You can find more details on the issue.

## Related Tickets & Documents
https://github.com/forem/internalEngineering/issues/276

## QA Instructions, Screenshots, Recordings

This may be a strange way to test, but it seemed like the easiest way to mimic a failure:
1. Get the last data update script `d = DataUpdateScript.last`
2. Take note of the filename
3. Destroy the data script `d.destroy`
4. Go to the actual file in your editor using the filename from step 2. It will probably look like this: 
```
module DataUpdateScripts
  class ResaveArticlesForCodeSnippetFullscreenIcon
    def run
      # We need to regenerate the markdown for code snippets
      # See https://github.com/forem/forem/issues/11747#issuecomment-753620156
      Article.published.find_each(&:save)
    end
  end
end
```
and change it in such a way that it fails. You can try  `Article.published.find_each(&:something)`
5. Re-run the data update scripts `rails data_updates:run`
6. Now look at the last data update script again by running `reload!` and `DataUpdateScript.last`
7. You should see that the script failed with an error which gets saved in the error field 
```
#<DataUpdateScript:0x00007f822b3bd648
 id: 1406,
 file_name: "20210104170542_resave_articles_for_code_snippet_fullscreen_icon",
 status: "failed",
 run_at: Mon, 18 Jan 2021 16:22:17 UTC +00:00,
 finished_at: Mon, 18 Jan 2021 16:22:17 UTC +00:00,
 created_at: Mon, 18 Jan 2021 16:22:17 UTC +00:00,
 updated_at: Mon, 18 Jan 2021 16:22:17 UTC +00:00,
 error: "undefined method `dosomething' for #<Article:0x00007ff39a3ef1d0>">
```
8. Also go to http://localhost:3000/admin/data_update_scripts and find the row corresponding to the data update script you just caused to fail. You should also see failed with the failure like below:
<img width="1680" alt="Screenshot 2021-01-20 at 13 03 55" src="https://user-images.githubusercontent.com/2786819/105166273-f657e000-5b1f-11eb-842e-4860ba85944f.png">

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

      have not been included_

## Added to documentation?

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [x] No documentation as yet

## [optional] Are there any post deployment tasks we need to perform?
No 

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/8L0Pky6C83SzkzU55a/giphy.gif)
